README.md
FFI::Sweet

Borrowing heavily from Ruby's FFI, let's try to make libffi and FFI::Raw easier to swallow.

Built in the interest of Audio::PortAudio::FFI

# References
* Perl FFI::Raw
    * https://metacpan.org/release/FFI-Raw
    * https://github.com/ghedo/p5-FFI-Raw
    * https://metacpan.org/requires/module/FFI::Raw
* libffi
    * https://github.com/atgreen/libffi
    * https://sourceware.org/libffi/
    * [Blog on using libffi from C](http://eli.thegreenplace.net/2013/03/04/flexible-runtime-interface-to-shared-libraries-with-libffi/)
* Ruby's Portaudio via ffi
    * https://github.com/jvoorhis/ruby-portaudio
* Ruby's FFI
    * https://github.com/ffi/ffi/wiki
    * https://github.com/ffi/ffi
    * https://groups.google.com/forum/#!forum/ruby-ffi
* Python's ctypes (on libffi)
    * http://docs.python.org/2/library/ctypes.html
    * [Blog on using ctypes](http://eli.thegreenplace.net/2013/03/09/python-ffi-with-ctypes-and-cffi/)
* Audio::PortAudio::FFI
    * https://github.com/awwaiid/Audio-PortAudio-FFI


Changes
Changelog for Compress-LZW

0.03 2014-01-23T00:28:05Z
 - Doc updates, add a sanity check, pile more junk on dzil (Meredith
   Howard)
 - Add Type::Tiny constraints for ~public attributes (Meredith Howard)
 - Add subtest for extended code size (Meredith Howard)
 - Add subtest for non-default init code size (Meredith Howard)
 - add tests for some non-default options and fixes to match (block resets
   and msb0 mode) (Meredith Howard)
 - Fix defaults included in method headings (Meredith Howard)
 - Fix hex constant (only failed to compile on some systems?) (Meredith
   Howard)

0.02 2014-01-15T01:13:03Z
 - Rearrange a bit and fill in more pod (Meredith Howard)
 - enable github, git stuff (Meredith Howard)
 - build changes (Meredith Howard)
 - Remove readme (Meredith Howard)
 - Add reset methods for object reuse, fixes for tests, pod stubs (Meredith
   Howard)
 - Expand tests (Meredith Howard)
 - Break out modules into separate files (Meredith Howard)
 - Clean up (Meredith Howard)
 - forget Changes (Meredith Howard)
 - Forget Changes, tweak dist.ini more (Meredith Howard)
 - dist.ini updates, add Changes file (Meredith Howard)
 - rearrange pod (Meredith Howard)
 - Implement decompression, add test for larger data that requires scaling
   up to 16 bits (Meredith Howard)
 - Add init_code_size, reorganize (Meredith Howard)
 - Rewrite compression (Meredith Howard)
 - Rewrite compression, modify tests. (Meredith Howard)
 - Initial (Meredith Howard)



lib/Compress/LZW.pm
package Compress::LZW;
# ABSTRACT: Pure-Perl implementation of scaling LZW

=head1 SYNOPSIS

 use Compress::LZW;
  
 my $compressed = compress($some_data);
 my $data       = decompress($compressed);
  
=head1 DESCRIPTION

C<Compress::LZW> is a perl implementation of the Lempel-Ziv-Welch compression
algorithm, which should no longer be patented worldwide.  It is shooting for
loose compatibility with the flavor of LZW found in the classic UNIX
compress(1), though there are a few variations out there today.  I test against
ncompress on Linux x86.

=cut

use strictures;

use base 'Exporter';

BEGIN {
  our @EXPORT      = qw/compress decompress/;
  our @EXPORT_OK   = qw(
    $MAGIC       $MASK_BITS    $MASK_BLOCK
    $RESET_CODE  $BL_INIT_CODE $NR_INIT_CODE
    $INIT_CODE_SIZE
  );
  our %EXPORT_TAGS = (
    const => \@EXPORT_OK,
  );
}

our $MAGIC          = "\037\235";
our $MASK_BITS      = 0x1f;
our $MASK_BLOCK     = 0x80;
our $RESET_CODE     = 256;
our $BL_INIT_CODE   = 257;
our $NR_INIT_CODE   = 256;
our $INIT_CODE_SIZE = 9;

use Compress::LZW::Compressor;
use Compress::LZW::Decompressor;

=func compress

Accepts a scalar, returns compressed data in a scalar.

Wraps L<Compress::LZW::Compressor>

=cut

sub compress {
  my ( $str ) = @_;
  
  return Compress::LZW::Compressor->new()->compress( $str );
}

=func decompress

Accepts a (compressed) scalar, returns decompressed data in a scalar.

Wraps L<Compress::LZW::Decompressor>

=cut


sub decompress {
  my ( $str ) = @_;
  
  return Compress::LZW::Decompressor->new()->decompress( $str );
}  

=head1 EXPORTS

Default: C<compress> C<decompress>

=head1 SEE ALSO

The implementations, L<Compress::LZW::Compressor> and
L<Compress::LZW::Decompressor>.

Other Compress::* modules, especially Compress::LZV1, Compress::LZF and
Compress::Zlib.

I definitely studied some other implementations that deserve credit, in
particular: Sean O'Rourke, E<lt>SEANOE<gt> - Original author,
C<Compress::SelfExtracting>, and another by Rocco Caputo which was posted
online.

=cut

1;


lib/Compress/LZW/Decompressor.pm
package Compress::LZW::Decompressor;
# ABSTRACT: Scaling LZW decompressor class

=head1 SYNOPSIS

 use Compress::LZW::Decompressor;
  
 my $d    = Compress::LZW::Decompressor->new();
 my $orig = $d->decompress( $lzw );
  
=cut

use Compress::LZW qw(:const);

use Types::Standard qw( Bool Int );

use Moo;
use namespace::clean;

=method decompress ( $input )

Decompress $input with the current settings and returns the result.

=cut

sub decompress {
  my $self = shift;
  my ( $data ) = @_;

  $self->reset;

  $self->{data}     = \$data;
  $self->{data_pos} = 0;

  $self->_read_magic;
  $self->{data_pos} = 24;

  $self->_str_reset;

  my $next_increase = 2 ** $self->{code_size};

  my $seen = $self->_read_code;
  my $buf  = $self->{str_table}{$seen};

  while ( defined( my $code = $self->_read_code ) ){

    if ( $self->{block_mode} and $code == $RESET_CODE ){
      # warn sprintf('reset table (%s, %s) at %s', $seen, $code, $self->{data_pos} - $self->{code_size});
      #reset table, next code, and code size
      $self->_str_reset;
      $next_increase = 2 ** $self->{code_size};

      $seen = $self->_read_code;
      $buf .= $self->{str_table}{$seen};
      
      next;
    }
    
    if ( defined ( my $word = $self->{str_table}{ $code } ) ){

      $buf .= $word;

      $self->{str_table}{ $self->{next_code} } = $self->{str_table}{ $seen } . substr($word,0,1);

    }
    elsif ( $code == $self->{next_code} ){
      
      my $word = $self->{str_table}{$seen};
           
      $self->{str_table}{$code} = $word . substr( $word, 0, 1 );
      
      $buf .= $self->{str_table}{$code};

    }
    else {
      die "($code != ". $self->{next_code} . ") input may be corrupt before bit $self->{data_pos}";
    }

    $seen = $code;
    
    # if next code expected will require a larger bit size
    if ( $self->{next_code} + 1 >= $next_increase ){
      if ( $self->{code_size} < $self->{max_code_size} ){
        # warn "decode up to $self->{code_size} bits at bit $self->{data_pos}";
        $self->{code_size} += 1;
        $next_increase     *= 2;
      }
      else {
        $self->{at_max_code} = 1;
      }
    }

    if ( $self->{at_max_code} == 0 ){
      $self->{next_code} += 1;
    }
    
  }
  return $buf;
}

=method reset ()

Resets the decompressor state for another round of input. Automatically
called at the beginning of ->decompress.

Resets the following internal state: code table, next code number, code
size, output buffer

=cut

sub reset {
  my $self = shift;
 
  $self->{data}        = undef;
  $self->{data_pos}    = 0;

  $self->_str_reset;
}

sub _str_reset {
  my $self = shift;
  
  $self->{str_table} = {
    map { $_ => chr($_) } 0 .. 255
  };
  
  $self->{code_size}   = $INIT_CODE_SIZE;
  $self->{next_code}   = $self->{block_mode} ? $BL_INIT_CODE : $NR_INIT_CODE;
  $self->{at_max_code} = 0;
}

sub _read_magic {
  my $self = shift;
  
  my $magic = substr( ${ $self->{data} }, 0, 3 );

  if ( length($magic) != 3 or substr($magic,0, 2) ne $MAGIC ){
    die "Invalid compress(1) header";
  }

  my $bits = ord( substr( $magic, 2, 1 ) );

  $self->{max_code_size} = $bits & $MASK_BITS;
  $self->{block_mode}    = ( $bits & $MASK_BLOCK ) >> 7;
}

sub _read_code {
  my $self = shift;

  if ( $self->{data_pos} > length( ${$self->{data}} ) * 8 ){
    # warn "bailing at $self->{data_pos} + $self->{code_size} > " . length( ${$self->{data}} ) *8;
    return undef;
  }
  
  my $code = 0;
  for ( 0 .. ($self->{code_size} - 1) ){
    $code |=
      vec( ${$self->{data}}, $self->{data_pos} + $_ , 1) << $_;
  }
  
  $self->{data_pos} += $self->{code_size};
  
  return undef if $code == 0 and $self->{data_pos} > length( ${$self->{data}} ) * 8;
  return $code;
  
}

1;


lib/Compress/LZW/Compressor.pm
package Compress::LZW::Compressor;
# ABSTRACT: Scaling LZW compressor class

=head1 SYNOPSIS

 use Compress::LZW::Compressor;
  
 my $c   = Compress::LZW::Compressor->new();
 my $lzw = $c->compress( $some_data );
  
=cut


use Compress::LZW qw(:const);

use Types::Standard qw( Bool Int );

use bytes;

use Moo;
use namespace::clean;

my $CHECKPOINT_BITS = 10_000;

=attr block_mode

Default: 1

Block mode is a feature added to LZW by compress(1). Once the maximum code size
has been reached, if the compression ratio falls (NYI) the code table and code
size can be reset, and a code indicating this reset is embedded in the output
stream.

May be 0 or 1.

=cut

has block_mode => (
  is      => 'ro',
  default => 1,
  isa     => Bool,
);

=attr max_code_size

Default: 16

Maximum size in bits that code output may scale up to.  This value is stored in
byte 3 of the compressed output so the decompressor can also stop at the same
size automatically.  Maximum code size implies a maximum code table size of C<2
** max_code_size>, which can be emptied and restarted mid-stream in
L</block_mode>.

May be between 9 and 31, inclusive.  The default of 16 is the largest supported
by compress(1), but Compress::LZW can handle up to 31 bits.

=cut

has max_code_size => ( # max bits
  is      => 'ro',
  default => 16,
  isa     => Type::Tiny->new(
    parent     => Int,
    constraint => sub { $_ >= $INIT_CODE_SIZE and $_ < $MASK_BITS },
    message    => sub { "$_ isn't between $INIT_CODE_SIZE and $MASK_BITS" },
  ),
);


=method compress ( $input )

Compresses $input with the current settings and returns the result.

=cut

sub compress {
  my $self = shift;
  my ( $str ) = @_;
  
  $self->reset;
  
  my $bytes_in;
  my ( $checkpoint, $last_ratio ) = ( 0, 0 );
 
  my $seen = '';

  for ( 0 .. length($str) ){
    my $char = substr($str, $_, 1);

    $bytes_in += 1;
    
    if ( exists $self->{code_table}{ $seen . $char } ){
      $seen .= $char;
    }
    else {      
      $self->_buf_write( $self->{code_table}{ $seen } );
      
      $self->_new_code( $seen . $char );
      
      $seen = $char;

      if ( $self->{at_max_code} and $self->block_mode ){
        if ( ! defined $checkpoint ){
          $checkpoint = $self->{buf_pos} + $CHECKPOINT_BITS;
        }
        elsif ( $bytes_in > $checkpoint ){
          my $ratio   = $bytes_in / ( $self->{buf_pos} / 8 );
          $last_ratio = 0 if !defined $last_ratio;

          
          if ( $ratio >= $last_ratio ){
            $last_ratio = $ratio;
            $checkpoint = $self->{buf_pos} + $CHECKPOINT_BITS;
          }
          elsif ( $ratio < $last_ratio ){
            # warn "Resetting code table ( $ratio < $last_ratio :: $self->{buf_pos} )";
            $self->_buf_write( $RESET_CODE );
            $self->_code_reset;

            undef $checkpoint;
            undef $last_ratio;
          }
        }
      }

    }
  }

  $self->_buf_write( $self->{code_table}{ $seen } );  #last bit of input
  # warn "final ratio: " . ($bytes_in / ($self->{buf_pos} / 8));
  
  return $self->{buf};
}


=method reset ()

Resets the compressor state for another round of compression. Automatically
called at the beginning of compress().

Resets the following internal state: Code table, next code number, code size,
output buffer, buffer position

=cut

sub reset {
  my $self = shift;
  
  # replace buf with empty buffer after magic bytes
  $self->{buf}     = $MAGIC
    . chr( $self->max_code_size | ( $self->block_mode ? $MASK_BLOCK : 0 ) );

  $self->{buf_pos} = length($self->{buf}) * 8;
  
  $self->_code_reset;
}


sub _code_reset {
  my $self = shift;
  
  $self->{code_table} = {
    map { chr($_) => $_ } 0 .. 255
  };

  $self->{at_max_code}   = 0;
  $self->{code_size}     = $INIT_CODE_SIZE;
  $self->{next_code}     = $self->block_mode ? $BL_INIT_CODE : $NR_INIT_CODE;
  $self->{next_increase} = 2 ** $self->{code_size};

}

sub _new_code {
  my $self = shift;
  my ( $word ) = @_;

  if ( $self->{next_code} >= $self->{next_increase} ){

    if ( $self->{code_size} < $self->{max_code_size} ){
      $self->{code_size}     += 1;
      $self->{next_increase} *= 2;
    }
    else {
      $self->{at_max_code} = 1;
    }
  }
  
  if ( $self->{at_max_code} == 0 ){
    $self->{code_table}{ $word } = $self->{next_code};
    $self->{next_code} += 1;
  }

}

sub _buf_write {
  my $self = shift;
  my ( $code ) = @_;

  return unless defined $code;
  
  if ( $code > ( 2 ** $self->{code_size} ) ){
    die "Code value $code too high for current code size $self->{code_size}";
  }

  my $wpos = $self->{buf_pos};
  # if ( $code == $RESET_CODE ){
  #   warn "wrote a reset code ($RESET_CODE) at $wpos";
  # }
  #~ warn "write $code \tat $code_size bits\toffset $wpos (byte ".int($wpos/8) . ')';
  
  if ( $code == 1 ){
    vec( $self->{buf}, $wpos, 1 ) = 1;
  }
  else {
    for my $bit ( 0 .. ($self->{code_size} - 1) ){
      
      if ( ($code >> $bit) & 1 ){
        vec( $self->{buf}, $wpos + $bit, 1 ) = 1;
      }
    }
  }
  
  $self->{buf_pos} += $self->{code_size};
}

1;
